python/ray/serve/_private/application_state.py

```diff
@@ -601,20 +601,26 @@ def _determine_app_status(self) -> Tuple[ApplicationStatus, str]:
         # application status to set.
         deployment_statuses = self.get_deployments_statuses()
         lowest_rank_status = min(deployment_statuses, key=lambda info: info.rank)
-        if lowest_rank_status.status == DeploymentStatus.UNHEALTHY:
+        if lowest_rank_status.status == DeploymentStatus.DEPLOY_FAILED:
```

Comment:
Do we also need to check `self._status` or add an assertion for it?

Reply:
I think that should be covered now by the logic in DeploymentStatus state transitions.

Reply:
hmm ok, I'm thinking something like if the app is already unhealthy, then we try to redeploy, and then any of the deployment have failed, I guess that should still be marked as `DEPLOY_FAILED` instead of just straight out `UNHEALTHY`. Feel like there could be a lot of those cases which I'm not sure if there are already test coverage and what's the expected behavior. Would suggest to add some tests to cover them, if not already, but can be a follow up in the future. I guess the existing app status is independent of the deployment status given that any of the deployment failed to deploy 👍